vllm/core/block_manager_v2.py

```diff
@@ -227,17 +228,63 @@ def fork(self, parent_seq: Sequence, child_seq: Sequence) -> None:
 
     def can_swap_in(self, seq_group: SequenceGroup,
                     num_lookahead_slots: int) -> bool:
-        return False
+        """
+        We go through all sequence in seq group to get their number of blocks 
+        touched and sum them up to see whether there is enough memory to swap in
+        """
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.SWAPPED):
+            block_table = self.block_tables[seq.seq_id]
+            num_touched_blocks += (
+                block_table.get_num_blocks_touched_by_append_slots(
+                    token_ids=seq.get_token_ids(),
+                    num_lookahead_slots=num_lookahead_slots,
+                ))
+        num_free_blocks = self.block_allocator.get_num_free_blocks(Device.GPU)
+        return num_free_blocks - num_touched_blocks >= self.watermark_blocks
```

Comment:
This looks pretty clean. Two thoughts:
* Can we find a good helper method API that serves both `can_swap_in` and `can_append_slots`?
* This will overestimate the number of blocks required by O(seq_len*best_of) in beam search + swap use-cases. Can we think of a way that doesn't suffer from such overallocation? It might require a new API in the block table itself.